records/README.md

```diff
@@ -0,0 +1,244 @@
+
+
+## Constraints
+
+IPRS has the following _hard_ constraints:
+
+- **MUST** be transport agnostic. (transport refers to how computers communicate).
+- **MUST** be replication agnostic. (replication refers to the protocol computers use to transfer and propagate whole records and other objects)
+- **MUST** be validity scheme agnostic. (validity scheme includes PKI, TI, and other "agreed upon" trusted infrastructure)
+- **MUST** be trustless: no trusted third parties are imposed by IPRS (though some may be adopted by a validity scheme. e.g. root CAs in the CA system PKI, or a blockchain in a blockchain TI). In most cases, users may have to trust each other (as they must trust the record value -- e.g. DNS), but in some cases there may be cryptographic schemes that enable full trustlessness.
+
+
+It is easy to be agnostic to transport, replication, and validity scheme as long as users can expect to control or agree upon the programs or protocols used in concert with IPRS. Concretely, the user can select specific transports or validity schemes to suit the user's application constraints. It is the user's responsibility to ensure both record crafters and verifiers agree upon these selections.
+
+## Construction
+
+### The Objects
+
+IPRS records are expressed as [merkledag](../merkledag) objects. This means that the records are linked authenticated data structures, and can be natively replicated over IPFS itself and other merkledag distribution systems.
+
+The objects:
+
+- A `Record` object expresses a value, a validity scheme, and validity data.
+- A `Signature` object could be used to sign and authenticate a record.
+- An `Encryption` object could be used to encrypt a record.
+
+```go
+Record Node {
+  Scheme   Link // link to a validity scheme
+  Value    Link // link to an object representing the value.
+  Validity Data // data needed to satisfy the validity scheme
+}
+```
+
+To achieve good performance, record storage and transfer should bundle all the necessary objects and transmit them together. While "the record object" is only one of the dag objects, "the full record" means a bundle of all objects needed to fully represent, verify, and use the record. (This recommendation does not necessarily include data that records _describe_, for example an ipfs provider record (which signals to consumers that certain data is available) would not include the data itself as part of "the full record").
+
+### The Interface
+
+The IPRS interface is below. It has a few types and functions. We use the Go language to express it, but this is language agnostic.
+
+```go
+// Record is the base type. user can define other types that
+// extend Record.
+type Record struct {
+  Scheme    Link // link to the validity scheme
+  Signature Link // link to a cryptographic signature over the rest of record
+  Value     Data // an opaque value
```

Comment:
On Wed, Jun 03, 2015 at 06:46:24PM -0700, Jeromy Johnson wrote:
> > +type Record struct {
> > +  Scheme    Link // link to the validity scheme
> > +  Signature Link // link to a cryptographic signature over the rest of record
> > +  Value     Data // an opaque value
> 
> Another thing, is the signature separate from the validity data? I'm
> okay with that, just want to make the distinction clear.

Just to keep the vocabulary consistent, @jbenet was putting the
signature under “correctness” not “validity” (since it's a
done-right-at-craft-time issue [1]), although they both fall under the
“validity scheme” [2].

I agree with [3] that it makes sense to allow a given validity scheme
to place correctness/validity information in additional links and/or
the data block as it sees fit, so long as it doesn't clobber the
base-Record "Scheme" or "Value" names (and do those have to be
title-cased?).

As it's currently written up, the linking for a signature is going to
be weird, since what you really want signed is *the record itself*.
The whole “signable part” business reminds me of OpenPGP with it's
“unhashed subpacket data” [4].  I played around with an OpenPGP
implementation while trying to rotate my main GnuPG key [5], and the
whole “unhashed data” business is a pain in the neck.  Can we just
make signature objects first-class citizens and allow (require?) folks
to push signatures instead of records.  Then the signature could link
to the record, which would carry the validity information and a link
to the record payload.

[1]: https://github.com/ipfs/specs/pull/7/files#r31354268
[2]: https://github.com/ipfs/specs/pull/7/files#diff-0cb9903b60857948374119aa5ca7a097R37
[3]: https://github.com/ipfs/specs/pull/7/files#r31687180
[4]: https://tools.ietf.org/html/rfc4880#section-5.2.3
[5]: http://git.tremily.us/?p=gpg-migrate.git